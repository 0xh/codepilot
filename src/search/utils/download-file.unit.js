import path from 'path'
import fs from 'graceful-fs'
import createTmpDirectory from './create-tmp-directory'
import downloadFile from './download-file'

require('jest-playback').setup(__dirname)

describe('@search/utils/download-file', () => {
  let testFilePath
  beforeEach(done => {
    createTmpDirectory('test').subscribe({
      next: dir => {
        testFilePath = path.join(dir, 'test')
        done()
      }
    })
  })

  it('warns when it fails to download a file', done => {
    const consoleWarn = jest
      .spyOn(console, 'warn')
      .mockImplementation(error => {
        try {
          expect(error.message).toMatch('Error downloading file')
        } catch (error) {
          done.fail(error)
        }
      })
    downloadFile('invalid-url', testFilePath).subscribe({
      complete() {
        try {
          expect(consoleWarn).toHaveBeenCalled()
        } catch (error) {
          done.fail(error)
        }
        consoleWarn.mockRestore()
        done()
      }
    })
  })

  it('emits an error when it fails to download a file with failOnError', () => {
    downloadFile('invalid-url', testFilePath, {
      emitError: true
    }).subscribe({
      error(error) {
        expect(error.message).toMatch('Error downloading file')
      }
    })
  })

  it('successfully downloads a file with a valid URL', done => {
    const next = jest.fn(downloadPath => {
      try {
        expect(downloadPath).toEqual(testFilePath)
        expect(fs.existsSync(downloadPath)).toEqual(true)
        expect(
          fs.readFileSync(downloadPath, { encoding: 'utf8' })
        ).toMatchSnapshot()
      } catch (error) {
        done.fail(error)
      }
    })
    downloadFile('https://jsonplaceholder.typicode.com/todos/1', testFilePath, {
      emitError: true
    }).subscribe({
      next,
      complete() {
        try {
          expect(next).toHaveBeenCalledTimes(1)
        } catch (error) {
          done.fail(error)
        }
        done()
      }
    })
  })
})
