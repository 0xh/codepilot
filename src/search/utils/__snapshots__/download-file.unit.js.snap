// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@search/utils/download-file successfully downloads a file with a valid URL 1`] = `
"{
  \\"userId\\": 1,
  \\"id\\": 1,
  \\"title\\": \\"delectus aut autem\\",
  \\"completed\\": false
}"
`;
